ultralytics/models/yolo/segment/val.py

```diff
@@ -97,6 +104,67 @@ def _prepare_pred(self, pred, pbatch, proto):
         pred_masks = self.process(proto, pred[:, 6:], pred[:, :4], shape=pbatch["imgsz"])
         return predn, pred_masks
 
+    def merge_masks_and_calculate_IoU(
+        self,
+        gt_masks: torch.Tensor,
+        pred_masks: torch.Tensor,
+        gt_cls: torch.Tensor,
+        pred_cls: torch.Tensor,
+        overlap=False,
+    ):
+        """
+        Merge ground truth masks and predicted masks by class and calculate IoU.
+
+        Args:
+            gt_masks (torch.Tensor): Ground truth masks.
+            pred_masks (torch.Tensor): Predicted masks.
+            gt_cls (torch.Tensor): Ground truth class labels.
+            pred_cls (torch.Tensor): Predicted class labels.
+            overlap (bool, optional): Flag indicating whether the gt masks overlap. Defaults to False.
+
+        Returns:
+            torch.Tensor, torch.Tensor, torch.Tensor: Unique ground truth class labels, unique predicted class labels,
+            and IoU scores.
+        """
+        if overlap:
+            nl = len(gt_cls)
+            index = torch.arange(nl, device=gt_masks.device).view(nl, 1, 1) + 1
+            gt_masks = gt_masks.repeat(nl, 1, 1)  # shape(1,640,640) -> (n,640,640)
+            gt_masks = torch.where(gt_masks == index, 1.0, 0.0)
+        if gt_masks.shape[1:] != pred_masks.shape[1:]:
+            gt_masks = F.interpolate(gt_masks[None], pred_masks.shape[1:], mode="bilinear", align_corners=False)[0]
+            gt_masks = gt_masks.gt_(0.5)
+
+        unique_gt_cls = gt_cls.unique().view(-1)
+        binary_gt_masks = torch.stack([gt_masks[gt_cls == c].sum(0).clamp_(0, 1) for c in unique_gt_cls], dim=0)
```

Comment:
@manhcuong02 one question here, I saw you are superposing all the instances of each class in one imagebefore calculating iou, can we directly calculate the iou between each instances? 
For example current way to get final miou is:
1. superpose all the instances of each classes in one image into one instance, for example two `person` in one image would be fused to one instance then calculate iou.
2. and to way to get the final miou is: `iou.sum()` / `len(instances)`, and the `len(instances)` is sum of the fused instances from step 1

Can we just calculate the iou in the original instances level and get final `miou`? i.e the formula of calculating final miou would be the same but eliminating the "superposition" step? 
The benefit of this is that we can reuse some code to implement miou calculation, without introducing more code to maintain

Reply:
@Laughing-q Thank you for your response. In the first version of my code, I utilized your `_process_batch` function to calculate mIoU on each instance. However, I encountered instances where I received mIoU values greater than 1 (`mIoU > 1`), which is incorrect. I suspect this might be due to overlapping predicted results in some cases (I've tested and observed such cases). Hence, I think calculating mIoU on each instance may not be correct. Have you encountered this issue before? Was it because of my previous incorrect calculations? Thank you sincerely